### PR TITLE
Optimism Dispute Game Support

### PR DESCRIPTION
**Pull request by Chomtana**
An improved approach to #36 

In this version, there's no need to split the code into two different files. We first try to find the index using the dispute game. If the dispute game is unavailable, we will fall back to using L2OutputOracle.

The function that needs L2OutputOracle has been replaced with OptimismPortalProxy, which is the core of the OP Stack. OptimismPortalProxy can derive both L2OutputOracle and DisputeGameFactory.

This implementation has been tested and works well.
* **Live demo:** https://app.ens.domains/disputegame.eth?tab=records
* **Gateway:** https://op-gateway-worker.optidomains.workers.dev
* **L1 Resolver:** [0xA3C417AcCe0243d9CC8006617A74217C2a696dB3](https://sepolia.etherscan.io/address/0xA3C417AcCe0243d9CC8006617A74217C2a696dB3)

image

> The deployed gateway implementation differs slightly from the one in the repository, with the ENS Tracker part removed and the app variable moved into a local variable that reinitializes on every request.

### CODE REVIEW

File: op-verifier/contracts/OPVerifier.sol [hunk 1]
```diff
@@ -1,43 +1,177 @@
...
-import { RLPReader } from "@eth-optimism/contracts-bedrock/src/libraries/rlp/RLPReader.sol";
-import { StateProof, EVMProofHelper } from "@ensdomains/evm-verifier/contracts/EVMProofHelper.sol";
-import { Types } from "@eth-optimism/contracts-bedrock/src/libraries/Types.sol";
-import { Hashing } from "@eth-optimism/contracts-bedrock/src/libraries/Hashing.sol";
+import {IEVMVerifier} from '@ensdomains/evm-verifier/contracts/IEVMVerifier.sol';
+import {RLPReader} from '@eth-optimism/contracts-bedrock/src/libraries/rlp/RLPReader.sol';
+import {StateProof, EVMProofHelper} from '@ensdomains/evm-verifier/contracts/EVMProofHelper.sol';
+import {Types} from '@eth-optimism/contracts-bedrock/src/libraries/Types.sol';
+import {Hashing} from '@eth-optimism/contracts-bedrock/src/libraries/Hashing.sol';
+import './interfaces/IDisputeGame.sol';
```
Comment (makoto) [inline op-verifier/contracts/OPVerifier.sol:9]: Can't we import from `@eth-optimism/contracts-bedrock` ? the file looks slightly from https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/src/dispute/interfaces/IDisputeGame.sol

File: op-verifier/contracts/OPVerifier.sol [hunk 2]
```diff
@@ -1,43 +1,177 @@
...
+import './interfaces/IDisputeGameFactory.sol';
```
Comment (makoto) [inline op-verifier/contracts/OPVerifier.sol:10]: Same for this

File: op-verifier/contracts/OPVerifier.sol [hunk 3]
```diff
@@ -1,43 +1,177 @@
...
 
 interface IL2OutputOracle {
-    function getL2Output(uint256 _l2OutputIndex) external view returns (Types.OutputProposal memory);
+    function getL2Output(
+        uint256 _l2OutputIndex
+    ) external view returns (Types.OutputProposal memory);
+}
+
+interface IOptimismPortal {
+    function l2Oracle() external view returns (IL2OutputOracle);
```
Comment (makoto) [inline op-verifier/contracts/OPVerifier.sol:30]: Can you point me in the source code of optimism repo where it's defined? I can't find the function on etherscan's verified code https://sepolia.etherscan.io/address/0x16Fc5058F25648194471939df75CF27A2fdC48BC#readProxyContract nor on their github repo https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/src/L1/OptimismPortal2.sol

File: op-verifier/contracts/OPVerifier.sol [hunk 4]
```diff
@@ -1,43 +1,177 @@
...
+        // If opOracle address is zero then revert with deprecated
+        if (address(opOracle) == address(0))
+            revert DeprecatedProof(OPWitnessProofType.L2OutputOracle, index);
+
+        // Get L2 Output Proposal at index
+        Types.OutputProposal memory l2out = opOracle.getL2Output(index);
+        return l2out.outputRoot;
+    }
+
+    function getDisputeGameOutput(
```
Comment (makoto) [inline op-verifier/contracts/OPVerifier.sol:91]: Are there any reference you used so that I can verify?

File: op-verifier/contracts/OPVerifier.sol (hunk 1 above)
Comment (Chomtana) [reply op-verifier/contracts/OPVerifier.sol:9]: I'm facing `Error HH411: The library src, imported from @eth-optimism/contracts-bedrock/src/dispute/interfaces/IDisputeGame.sol, is not installed. Try installing it using npm.` due to they have migrated to use foundry absolute import path.

I have asked at https://github.com/NomicFoundation/hardhat/discussions/5049

> I think this is an issue with the way Hardhat and Foundry resolve Solidity imports and they clash - Foundry resolution is based on remappings, while Hardhat has a treats direct imports as coming from the node_modules folder.
> 
> We are looking to add first class support for remappings to Hardhat to support these sorts of cases.

It sounds like they are implementing this feature but is not supported yet.

File: op-verifier/contracts/OPVerifier.sol (hunk 3 above)
Comment (Chomtana) [reply op-verifier/contracts/OPVerifier.sol:30]: Old contract and new contract have different method but always deployed to the same address (Upgradeable Proxy)

* OptimismPortal can resolve to l2Oracle only on the old contract
* OptimismPortal can resolve to disputeGameFactory only on the new contract (Because old method is replaced by spacer_54_0_20)

This makes it possible to automatically determine whether OP is using a new or old implementation

For the source code
* Old: https://github.com/ethereum-optimism/optimism/blob/2b3f00df3e865bd55369c5d6bf4a263839b8f0cc/packages/contracts-bedrock/src/L1/OptimismPortal.sol#L71
* New: https://github.com/ethereum-optimism/optimism/blob/2b3f00df3e865bd55369c5d6bf4a263839b8f0cc/packages/contracts-bedrock/src/L1/OptimismPortal2.sol#L80

Probably I should use try-catch. Will test on remix first to see what's the behaviour of calling an undefined function.

File: op-verifier/contracts/OPVerifier.sol (hunk 4 above)
Comment (Chomtana) [reply op-verifier/contracts/OPVerifier.sol:91]: Would be `proveWithdrawalTransaction`

https://github.com/ethereum-optimism/optimism/blob/2b3f00df3e865bd55369c5d6bf4a263839b8f0cc/packages/contracts-bedrock/src/L1/OptimismPortal2.sol#L254-L328

File: op-verifier/contracts/OPVerifier.sol (hunk 3 above)
Comment (makoto) [reply op-verifier/contracts/OPVerifier.sol:30]: I see. so your code supports pre-post migrations?

Comment (Chomtana) [reply op-verifier/contracts/OPVerifier.sol:30]: Yes, for OP Stack chain that hasn't upgraded

File: op-verifier/contracts/OPVerifier.sol (hunk 4 above)
Comment (makoto) [reply op-verifier/contracts/OPVerifier.sol:91]: How did you manage to test each failure scenario like  GameChallenged, GameTooEarly, DeprecatedProof or are they even possible to test?

File: op-verifier/contracts/OPVerifier.sol (hunk 1 above)
Comment (makoto) [reply op-verifier/contracts/OPVerifier.sol:9]: Did you also try to upgrade `@eth-optimism/contracts-bedrock` to the latest version?

File: op-gateway/src/OPProofService.ts [hunk 5]
```diff
@@ -15,58 +42,200 @@ const L2_OUTPUT_ORACLE_ABI = [
...
+          DISPUTE_GAME_FACTORY_ABI,
+          this.l1Provider
+        );
+      } catch (err) {
+        console.error(err)
+        return null;
+      }
+    }
+
+    if (this.disputeGameFactory) {
```
Comment (Arachnid) [inline op-gateway/src/OPProofService.ts:115]: Can this ever be false?

File: op-gateway/src/OPProofService.ts [hunk 6]
```diff
@@ -15,58 +42,200 @@ const L2_OUTPUT_ORACLE_ABI = [
...
+      let disputeGameIndex = -1;
+      let disputeGameLocalIndex = -1;
+
+      for (let i = 0; i < games.length; i++) {
+        if (timestampNow - Number(games[i].timestamp) >= this.delay) {
+          disputeGameIndex = Number(games[i].index);
+          disputeGameLocalIndex = i;
+          break;
+        }
+      }
```
Comment (Arachnid) [inline op-gateway/src/OPProofService.ts:143]: This should probably be a function to find the game to use.

File: op-gateway/src/OPProofService.ts [hunk 7]
```diff
@@ -15,58 +42,200 @@ const L2_OUTPUT_ORACLE_ABI = [
...
+
+      // If game is not found then fallback to the L2OutputOracle
+      // for just merged case
```
Comment (Arachnid) [inline op-gateway/src/OPProofService.ts:146]: This would be easier to read if it returns null if `disputeGameIndex == -1`, making it possible to remove the if block below.

File: op-gateway/src/OPProofService.ts [hunk 8]
```diff
@@ -15,58 +42,200 @@ const L2_OUTPUT_ORACLE_ABI = [
...
+          gameCount - 1,
+          50
+        );
+
+      const timestampNow = (await this.l2Provider.getBlock('latest'))!
+        .timestamp;
+      let disputeGameIndex = -1;
+      let disputeGameLocalIndex = -1;
+
+      for (let i = 0; i < games.length; i++) {
```
Comment (Arachnid) [inline op-gateway/src/OPProofService.ts:137]: This procedure will fail if we just didn't fetch enough games. Some kind of interpolation search might be the best approach here.

Caching is also going to be essential to reduce overhead; even an old cached result can be used as a lower bound for fetching a new one.

File: op-gateway/src/OPProofService.ts [hunk 9]
```diff
@@ -15,58 +42,200 @@ const L2_OUTPUT_ORACLE_ABI = [
...
+          L2_OUTPUT_ORACLE_ABI,
+          this.l1Provider
+        );
+      } catch (err) {
+        console.error(err)
+        return null;
+      }
+    }
+
+    if (this.l2OutputOracle) {
```
Comment (Arachnid) [inline op-gateway/src/OPProofService.ts:199]: Can this ever be false?

File: op-gateway/src/OPProofService.ts [hunk 10]
```diff
@@ -15,58 +42,200 @@ const L2_OUTPUT_ORACLE_ABI = [
...
+      /**
+       * Get the latest output from the L2Oracle. We're building the proof with this batch
+       * We go a few batches backwards to avoid errors like delays between nodes
+       *
+       */
+      const l2OutputIndex =
+        Number(await this.l2OutputOracle.latestOutputIndex()) - this.delay;
```
Comment (Arachnid) [inline op-gateway/src/OPProofService.ts:206]: This should also check against the timestamp rather than a number of blocks.

File: op-verifier/contracts/OPVerifier.sol [hunk 11]
```diff
@@ -1,43 +1,177 @@
...
+        ) = disputeGameFactory.gameAtIndex(index);
+        Claim outputRoot = gameProxy.rootClaim();
+
+        // Unwrap gameCreationTime to uint64
+        uint64 gameCreationTime;
+        assembly {
+            gameCreationTime := gameCreationTimeRaw
+        }
+
+        // Wait for challenger to challenge the dispute game
```
Comment (Arachnid) [inline op-verifier/contracts/OPVerifier.sol:118]: We also need a max age, to prevent gateways from returning old results.

File: op-verifier/contracts/OPVerifier.sol [hunk 12]
```diff
@@ -1,43 +1,177 @@
...
+                disputeGameMinAge
+            );
+        }
+
+        // Revert if gameType is not the one used for L2 withdrawal
+        if (gameType.raw() != respectedGameType.raw()) {
+            revert GameTypeMismatch(index, respectedGameType, gameType);
+        }
+
+        // Revert if the game is challenged
```
Comment (Arachnid) [inline op-verifier/contracts/OPVerifier.sol:132]: Can we revert even if there's a challenge initiated? Eg, only allow unchallenged or finalized games?

File: op-verifier/contracts/interfaces/IDisputeGame.sol [hunk 13]
```diff
@@ -0,0 +1,156 @@
+// SPDX-License-Identifier: MIT
```
Comment (Arachnid) [inline op-verifier/contracts/interfaces/IDisputeGame.sol:1]: This needs to include some information on where it's sourced from, as well as copyright holders.

File: op-verifier/contracts/interfaces/IDisputeGameFactory.sol [hunk 14]
```diff
@@ -0,0 +1,135 @@
+// SPDX-License-Identifier: MIT
```
Comment (Arachnid) [inline op-verifier/contracts/interfaces/IDisputeGameFactory.sol:1]: As above.

**Review by Arachnid — CHANGES_REQUESTED** [review]
Thanks again for writing this!

File: op-gateway/src/OPProofService.ts (hunk 5 above)
Comment (Chomtana) [reply op-gateway/src/OPProofService.ts:115]: It will be false if that chain is still using an old L2OutputOracle

File: op-gateway/src/OPProofService.ts (hunk 6 above)
Comment (Chomtana) [reply op-gateway/src/OPProofService.ts:143]: Yeah, we may need to move the challenged assertion part to the game finding to skip challenged malicious games.

File: op-verifier/contracts/OPVerifier.sol (hunk 12 above)
Comment (Chomtana) [reply op-verifier/contracts/OPVerifier.sol:132]: That will make it vulnerable to malicious challengers that try to challenge random game?

File: op-gateway/src/OPProofService.ts (hunk 9 above)
Comment (Chomtana) [reply op-gateway/src/OPProofService.ts:199]: It will be false if that chain has migrated to DisputeGame

File: op-gateway/src/OPProofService.ts (hunk 5 above)
Comment (Arachnid) [reply op-gateway/src/OPProofService.ts:115]: But the previous block either sets it or returns - so execution should never reach this point with `this.disputeGameFactory` being null or undefined.

File: op-gateway/src/OPProofService.ts [hunk 15]
```diff
@@ -15,58 +42,200 @@ const L2_OUTPUT_ORACLE_ABI = [
...
+    if (!this.disputeGameFactory) {
+      try {
+        const disputeGameFactoryAddress = await this.optimismPortal.disputeGameFactory();
+  
+        this.disputeGameFactory = new Contract(
+          disputeGameFactoryAddress,
+          DISPUTE_GAME_FACTORY_ABI,
+          this.l1Provider
+        );
+      } catch (err) {
```
Comment (Arachnid) [inline op-gateway/src/OPProofService.ts:109]: An error setting this address, or other issue, should be cached here to avoid trying to do it on every call. Perhaps this and the other contract could both be initialized in the constructor instead of doing it just-in-time?

File: op-gateway/src/OPProofService.ts (hunk 9 above)
Comment (Arachnid) [reply op-gateway/src/OPProofService.ts:199]: As above, the previous block either sets it or returns, so it shouldn't be possible to reach this line with the value unset.

**Review by Arachnid — CHANGES_REQUESTED** [review]

File: op-gateway/src/OPProofService.ts (hunk 8 above)
Comment (Chomtana) [reply op-gateway/src/OPProofService.ts:137]: We can also deploy our read only smart contract to handle this job with lower latency

Comment (Chomtana) [reply op-gateway/src/OPProofService.ts:137]: https://github.com/Opti-domains/dispute-game-lookup

This contract can handle the entire flow in one call. Will integrate in a few days

Deployed to sepolia at https://sepolia.etherscan.io/address/0x51385dc119e47f54fca98147107bc3f9d9e3b395

But can't verify contract due to import `src/...` bug haunted even on foundry project itself...

File: op-verifier/contracts/OPVerifier.sol (hunk 1 above)
Comment (Chomtana) [reply op-verifier/contracts/OPVerifier.sol]: I have build another NPM package and use foundry flatten on optimism interfaces

https://github.com/Opti-domains/dispute-game-lookup/blob/main/npm/contracts/interfaces/optimism/IDisputeGameFactory.sol

File: op-verifier/contracts/interfaces/IDisputeGame.sol (hunk 13 above)
Comment (Chomtana) [reply op-verifier/contracts/interfaces/IDisputeGame.sol]: I have build another NPM package and use foundry flatten on optimism interfaces. It should include information on where it's sourced from as per foundry standard.

https://github.com/Opti-domains/dispute-game-lookup/blob/main/npm/contracts/interfaces/optimism/IDisputeGameFactory.sol

File: op-gateway/src/OPProofService.ts (hunk 15 above)
Comment (Chomtana) [reply op-gateway/src/OPProofService.ts]: Using OPOutputLookup contract I have built at https://github.com/Opti-domains/dispute-game-lookup/blob/main/src/OPOutputLookup.sol to do the calculation and advanced search in a single call. So, the lowest latency one can acheive.

File: op-verifier/contracts/OPVerifier.sol (hunk 11 above)
Comment (Chomtana) [reply op-verifier/contracts/OPVerifier.sol]: Max age is implemented in OPOutputLookup contract I have built at https://github.com/Opti-domains/dispute-game-lookup/blob/main/src/OPOutputLookup.sol. So do L2OutputOracle now using min and max age.

File: op-gateway/src/OPProofService.ts (hunk 10 above)
Comment (Chomtana) [reply op-gateway/src/OPProofService.ts]: Max age is implemented in OPOutputLookup contract I have built at https://github.com/Opti-domains/dispute-game-lookup/blob/main/src/OPOutputLookup.sol. So do L2OutputOracle now using min and max age.

**Comment by Chomtana** [comment]
Merged and pushed the new version

Using OPOutputLookup contract I have built at https://github.com/Opti-domains/dispute-game-lookup/blob/main/src/OPOutputLookup.sol to do the calculation and advanced search in a single call. So, the lowest latency one can achieve.

Pending unit test on https://github.com/Opti-domains/dispute-game-lookup

However, it works manual testing: https://app.ens.domains/disputegame.eth?tab=records (On Sepolia testnet)

File: op-gateway/src/OPProofService.ts [hunk 16]
```diff
@@ -5,72 +5,99 @@ import {
...
-  l2OutputIndex: number;
+  proofType: OPWitnessProofType;
+  index: number;
 }
 
-const L2_OUTPUT_ORACLE_ABI = [
-  'function latestOutputIndex() external view returns (uint256)',
-  'function getL2Output(uint256 _l2OutputIndex) external view returns (tuple(bytes32 outputRoot, uint128 timestamp, uint128 l2BlockNumber))',
-];
+// OPOutputLookup contract is deployed with deterministic deployment
```
Comment (Arachnid) [inline op-gateway/src/OPProofService.ts:33]: Can you link to the code/repo here?

**Review by Arachnid — CHANGES_REQUESTED** [review]
Thanks! Can you include the source for the helper contract in this repo, and deploy it as part of the tests?

**Comment by Chomtana** [comment]
Hi @Arachnid @makoto 

Sorry for the delay, I have been busy for these weeks!

Helper contracts and libraries have been merged into the repo and a deployment script to deterministically deploy OPOutputLookup have been added.

Some trick has been implemented to be able to import `@eth-optimism/contracts-bedrock` into the contracts rather than copying the code.

Current contract and gateway (with app variable moved into a local variable) has been successfully tested on current implementation.

https://app.ens.domains/disputegame.eth?tab=records (Sepolia)

image

File: op-verifier/contracts/OPVerifier.sol (hunk 2 above)
Comment (Chomtana) [reply op-verifier/contracts/OPVerifier.sol]: Some trick has been implemented to be able to import `@eth-optimism/contracts-bedrock` into the contracts rather than copying the code.

File: op-gateway/src/OPProofService.ts (hunk 16 above)
Comment (Chomtana) [reply op-gateway/src/OPProofService.ts:33]: https://github.com/Opti-domains/evmgateway/blob/op-fault-proof-new/op-verifier/contracts/lib/OPOutputLookup.sol

File: op-gateway/src/OPProofService.ts [hunk 17]
```diff
@@ -5,72 +5,100 @@ import {
...
-     *       bytes32 outputRoot;
-     *       uint128 timestamp;
-     *       uint128 l2BlockNumber;
-     *      }
-     */
-    const outputProposal = await this.l2OutputOracle.getL2Output(l2OutputIndex);
+    const block = await this.opOutputLookup.getOPProvableBlock(
+      await this.optimismPortal.getAddress(),
+      this.minAge,
+      1000000000
```
Comment (Arachnid) [inline op-gateway/src/OPProofService.ts]: What parameter is this? Can you leave a comment on it?

File: op-verifier/contracts/lib/DisputeGameLookup.sol [hunk 18]
```diff
@@ -0,0 +1,366 @@
...
+        // Reject dispute game that has been expired
+        if (maxAge > 0 && block.timestamp - gameCreationTime > maxAge) {
+            revert GameExpired(
+                index,
+                block.timestamp - gameCreationTime,
+                maxAge
+            );
+        }
+
+        // Revert if the game is challenged
```
Comment (Arachnid) [inline op-verifier/contracts/lib/DisputeGameLookup.sol]: We should also reject games that have an open challenge, not just games that have already been successfully challenged. Otherwise an attacker can keep using an invalid state during the whole challenge process.

File: op-verifier/contracts/lib/DisputeGameLookup.sol [hunk 19]
```diff
@@ -0,0 +1,366 @@
...
+    }
+
+    /**
+     * @notice Finds the search bound for binary search based on the maximum timestamp.
+     * @param disputeGameFactory The dispute game factory.
+     * @param maxTimestamp The maximum timestamp to search.
+     * @return lo The lower bound index.
+     * @return hi The upper bound index.
+     */
+    function _findSearchBound(
```
Comment (Arachnid) [inline op-verifier/contracts/lib/DisputeGameLookup.sol]: These two binary searches are a clever approach when we don't know where to start!

I have an alternative suggestion that may prove simpler to implement and more efficient however: interpolation search. Start by fetching the timestamp of the oldest dispute game, and then take the ratio of `(now-maxTimestamp)/(now-oldestTimestamp)`. Use that to pick an intermediate element (instead of always picking the middle, as in binary search) and repeat as in binary search.

This should let us search the full space in one go, without spending a lot of time on values that are way too old as we would if we just did a binary search on the whole set of games.

It may help to set the midpoint to, say, 95% of what a naive evaluation would expect on the first iteration, to ensure the element we want usually falls within the smaller window.

**Review by Arachnid — CHANGES_REQUESTED** [review]
Thank you for your persistent hard work on this! I'd say it's very close to done.

I left a couple of comments; aside from them are you able to write some tests using a mock for the Optimism contracts so that we can verify the binary search works as expected?

File: op-verifier/contracts/lib/DisputeGameLookup.sol (hunk 19 above)
Comment (Chomtana) [reply op-verifier/contracts/lib/DisputeGameLookup.sol]: I have tried to implement this but it is very unstable compared to the two binary search approach

Like it doesn't work on some minimum time for an unknown reason in a random way

Comment (Chomtana) [reply op-verifier/contracts/lib/DisputeGameLookup.sol]: image

Old two binary search work on any high minimum time

image

Interpolation search don't work on high minimum time

File: op-verifier/contracts/lib/DisputeGameLookup.sol (hunk 18 above)
Comment (Chomtana) [reply op-verifier/contracts/lib/DisputeGameLookup.sol]: That would make it delay a few days at least if we check for the `IN_PROGRESS` status. I haven't seen a function to directly check for an open challenge.

Comment (Chomtana) [reply op-verifier/contracts/lib/DisputeGameLookup.sol]: Can be seen at https://sepolia.etherscan.io/address/0x99b0e2a750Cd985A3707a8C792BB0ADb28d148b4#readContract

That try to implement this one

image

image

It go back 4 days in OP sepolia testnet

File: op-verifier/README.md [hunk 20]
```diff
@@ -63,6 +63,17 @@ The tests will require small modifications to work on public testnets; specifica
 
 - https://base-sepolia-gateway-worker.ens-cf.workers.dev/{sender}/{data}.json
 
+## OPOutputLookup
```
Comment (Arachnid) [inline op-verifier/README.md:66]: Can you include the compilation parameters that were used here, so people can reproduce the build?

File: op-verifier/contracts/lib/DisputeGameLookup.sol (hunk 19 above)
Comment (Arachnid) [reply op-verifier/contracts/lib/DisputeGameLookup.sol]: Thanks for giving it a try! Did you try debugging it in Tenderly to see what's going wrong?

File: op-verifier/contracts/lib/DisputeGameLookup.sol (hunk 18 above)
Comment (Arachnid) [reply op-verifier/contracts/lib/DisputeGameLookup.sol]: I'm a bit confused by this. Reading the OP code, it looks like the assumption is that every dispute game in fact corresponds to an active dispute. Are dispute games automatically created for each new block committed, or are they only created when someone challenges a block?

**Review by Arachnid — CHANGES_REQUESTED** [review]

File: op-verifier/contracts/lib/DisputeGameLookup.sol (hunk 18 above)
Comment (clowestab) [reply op-verifier/contracts/lib/DisputeGameLookup.sol]: Dispute games are created for every state root proposal and the successful resolution of a game is what updates the anchor state in the AnchorStateRegistry.

In principle anyone can create a game (and Optimism do have confirmed production test canaries trying to break things with invalid root claims) but the game has to resolve in favour of the proposed state for the anchor to be updated.

https://github.com/ethereum-optimism/optimism/blob/c2f290d479f7319776205c2640f140a8423cb987/packages/contracts-bedrock/src/dispute/AnchorStateRegistry.sol#L59

Comment (Arachnid) [reply op-verifier/contracts/lib/DisputeGameLookup.sol]: Okay, so there's no state in which a proposed state root has simply not been challenged by anyone? Is there a way to see if anyone has actually begun the dispute process?

Comment (Chomtana) [reply op-verifier/contracts/lib/DisputeGameLookup.sol]: It may be possible to detect that the length of `claimData` is greater than 1, which means someone has tried to challenge, but I'm not 100% sure.

File: op-verifier/README.md (hunk 20 above)
Comment (Chomtana) [reply op-verifier/README.md:66]: Included

* Address: 0x956929bf186c26b4521aA31145948f4917f4D16e
* Deterministic Deployment Proxy: 0x4e59b44847b379578588920cA78FbF26c0B4956C
* Salt: 0x0000000000000000000000000000000000000000000000000000000000000000
* Contract: lib/OPOutputLookup.sol:OPOutputLookup
* Solidity: 0.8.26
* Optimizer Runs: 200

File: op-verifier/contracts/lib/DisputeGameLookup.sol (hunk 18 above)
Comment (Chomtana) [reply op-verifier/contracts/lib/DisputeGameLookup.sol]: Implemented open challenge detection with these checks

```solidity
    function _isGameChallenging(
        IChallengingDisputeGame proxy
    ) internal view returns (bool) {
        // If the game is finalized, we ignore the challenging status
        if (proxy.status() == GameStatus.DEFENDER_WINS) return false;

        // If claimData length is greater than 1, the game is challenging
        if (proxy.claimDataLen() > 1) {
            return true;
        }

        // If root claim is countered by someone then the game is challenging
        (, address counteredBy, , , , , ) = proxy.claimData(0);
        if (counteredBy != address(0)) return true;

        return false;
    }
```

And fixed interpolation search revert on case that lower bound has moved beyond the max timestamp

Comment (Arachnid) [reply op-verifier/contracts/lib/DisputeGameLookup.sol]: Looks good! Is interpolation search working fine now then? Some tests (and benchmarks) are all that's missing to merge this, I think.

Thank you again for all your hard work on this PR.

File: op-gateway/src/OPProofService.ts (hunk 17 above)
Comment (Chomtana) [reply op-gateway/src/OPProofService.ts]: It's Unlimited maximum age to have the gateway fetch output root regardless of maximum age parameter as it will be verified on the contract anyway. Commented

File: op-verifier/contracts/lib/DisputeGameLookup.sol (hunk 18 above)
Comment (Chomtana) [reply op-verifier/contracts/lib/DisputeGameLookup.sol]: Test cases have been implemented in my [dispute-game-lookup](https://github.com/Opti-domains/dispute-game-lookup/tree/main) repository. I am using Foundry's fuzz testing combined with mainnet forking to identify and fix edge cases, including integer underflow and rounding errors.

The test cases can be found at the following links:
* [DisputeGameLookup](https://github.com/Opti-domains/dispute-game-lookup/blob/main/test/DisputeGameLookup.t.sol)
* [L2OutputOracleLookup](https://github.com/Opti-domains/dispute-game-lookup/blob/main/test/L2OutputOracleLookup.t.sol)

The files in the [dispute-game-lookup](https://github.com/Opti-domains/dispute-game-lookup/tree/main) repo and this repo differ only in the import section, as shown in these diffs:
* [DisputeGameLookup](https://www.diffchecker.com/99DREhBY/)
* [L2OutputOracleLookup](https://www.diffchecker.com/66xKoWgd/)
* [OPOutputLookup](https://www.diffchecker.com/Tug5lgIh/)

Comment (Chomtana) [reply op-verifier/contracts/lib/DisputeGameLookup.sol]: For benchmarking, it take about 3.3s for each record processed in parallel on chrome network profiling

image

https://app.ens.domains/disputegame.eth (On Sepolia)

**Comment by GavinPacini** [comment]
Hey, just wondering what's required here to get this into `main`? It seems to be working but I'm happy to pick it up if you have more change requests @Arachnid.

**Comment by Chomtana** [comment]
This version is working, but there is still room for improvement, such as adding an anchor or backup feature for dispute game types.
